File: typecast-php/src/SpeechComposer.php
<?php

declare(strict_types=1);

namespace Neosapience\Typecast;

use Neosapience\Typecast\Models\Output;
use Neosapience\Typecast\Models\PresetPrompt;
use Neosapience\Typecast\Models\Prompt;
use Neosapience\Typecast\Models\SmartPrompt;
use Neosapience\Typecast\Models\TTSRequest;
use Neosapience\Typecast\Models\TTSResponse;

class SpeechPart
{
    public function __construct(
        public string $kind,
        public ?string $text = null,
        public ?float $seconds = null,
    ) {
    }
}

class ComposerSettings
{
    public function __construct(
        public ?string $voiceId = null,
        public ?string $model = null,
        public ?string $language = null,
        public Prompt|PresetPrompt|SmartPrompt|null $prompt = null,
        public ?Output $output = null,
        public ?int $seed = null,
    ) {
    }
}

class SpeechComposer
{
    private ComposerSettings $defaults;

    /** @var array<int, array<string, mixed>> */
    private array $parts = [];

    public function __construct(private readonly TypecastClient $client)
    {
        $this->defaults = new ComposerSettings();
    }

    public function defaults(ComposerSettings $settings): self
    {
        $this->defaults = self::mergeSettings($this->defaults, $settings);
        return $this;
    }

    public function say(string $text, ?ComposerSettings $overrides = null): self
    {
        $this->parts[] = [
            'kind' => 'speech',
            'text' => $text,
            'settings' => self::mergeSettings($this->defaults, $overrides ?? new ComposerSettings()),
        ];
        return $this;
    }

    /**
     * Inserts silence between speech segments.
     *
     * $seconds is a duration in seconds. Use 0.3 for 300 ms, 3 for three seconds.
     */
    public function pause(float $seconds): self
    {
        $this->parts[] = ['kind' => 'pause', 'seconds' => $seconds];
        return $this;
    }

    public function generate(): TTSResponse
    {
        $plan = $this->buildPlan();
        $hasSpeech = false;
        foreach ($plan as $part) {
            if ($part['kind'] === 'speech') {
                $hasSpeech = true;
                break;
            }
        }
        if (!$hasSpeech) {
            throw new \InvalidArgumentException('at least one speech segment is required');
        }

        $formats = [];
        foreach ($plan as $part) {
            if ($part['kind'] === 'speech' && $part['settings']->output?->audioFormat !== null) {
                $formats[$part['settings']->output->audioFormat] = true;
            }
        }
        if (count($formats) > 1) {
            throw new \InvalidArgumentException('composed speech segments must use one audio format');
        }
        $outputFormat = array_key_first($formats) ?? 'wav';
        $segments = [];
        foreach ($plan as $part) {
            if ($part['kind'] === 'pause') {
                $seconds = $part['seconds'];
                if (!self::isValidPause($seconds)) {
                    throw new \InvalidArgumentException('pause seconds must be greater than 0');
                }
                $segments[] = ['type' => 'pause', 'duration_seconds' => $seconds];
                continue;
            }
            $segments[] = ['type' => 'tts'] + self::requestFromPart($part, $outputFormat)->toArray();
        }
        return $this->client->composeTextToSpeech($segments);
    }

    /**
     * @return array<int, array<string, mixed>>
     */
    private function buildPlan(): array
    {
        $plan = [];
        foreach ($this->parts as $part) {
            if ($part['kind'] === 'pause') {
                if (!self::isValidPause($part['seconds'])) {
                    throw new \InvalidArgumentException('pause seconds must be greater than 0');
                }
                $plan[] = $part;
                continue;
            }

            foreach (parsePauseMarkup($part['text']) as $parsed) {
                if ($parsed->kind === 'pause') {
                    $plan[] = ['kind' => 'pause', 'seconds' => $parsed->seconds];
                    continue;
                }
                if (trim((string) $parsed->text) === '') {
                    continue;
                }
                /** @var ComposerSettings $settings */
                $settings = $part['settings'];
                if (trim((string) $settings->voiceId) === '') {
                    throw new \InvalidArgumentException('voiceId is required for composed speech segments');
                }
                if (trim((string) $settings->model) === '') {
                    throw new \InvalidArgumentException('model is required for composed speech segments');
                }
                $plan[] = [
                    'kind' => 'speech',
                    'text' => $parsed->text,
                    'settings' => $settings,
                ];
            }
        }
        return $plan;
    }

    private static function mergeSettings(ComposerSettings $base, ComposerSettings $override): ComposerSettings
    {
        return new ComposerSettings(
            voiceId: $override->voiceId ?? $base->voiceId,
            model: $override->model ?? $base->model,
            language: $override->language ?? $base->language,
            prompt: $override->prompt ?? $base->prompt,
            output: self::mergeOutput($base->output, $override->output),
            seed: $override->seed ?? $base->seed,
        );
    }

    private static function mergeOutput(?Output $base, ?Output $override): ?Output
    {
        if ($base === null && $override === null) {
            return null;
        }
        return new Output(
            volume: $override?->volume ?? $base?->volume,
            targetLufs: $override?->targetLufs ?? $base?->targetLufs,
            audioPitch: $override?->audioPitch ?? $base?->audioPitch,
            audioTempo: $override?->audioTempo ?? $base?->audioTempo,
            audioFormat: $override?->audioFormat ?? $base?->audioFormat,
        );
    }

    /**
     * @param array<string, mixed> $part
     */
    private static function requestFromPart(array $part, string $outputFormat = 'wav'): TTSRequest
    {
        /** @var ComposerSettings $settings */
        $settings = $part['settings'];
        return new TTSRequest(
            voiceId: (string) $settings->voiceId,
            text: (string) $part['text'],
            model: (string) $settings->model,
            language: $settings->language,
            prompt: $settings->prompt,
            output: self::mergeOutput($settings->output, new Output(volume: null, audioPitch: null, audioTempo: null, audioFormat: $outputFormat)),
            seed: $settings->seed,
        );
    }

    private static function isValidPause(float $seconds): bool
    {
        return is_finite($seconds) && $seconds > 0;
    }
}

/**
 * @return SpeechPart[]
 */
function parsePauseMarkup(string $text): array
{
    $parts = [];
    $lastEmit = 0;
    $searchFrom = 0;
    while (($tokenStart = strpos($text, '<|', $searchFrom)) !== false) {
        $bodyStart = $tokenStart + 2;
        $bodyEnd = strpos($text, '|>', $bodyStart);
        if ($bodyEnd === false) {
            break;
        }
        $tokenEnd = $bodyEnd + 2;
        $tokenBody = substr($text, $bodyStart, $bodyEnd - $bodyStart);
        if (str_ends_with($tokenBody, 's')) {
            $secondsText = substr($tokenBody, 0, -1);
            if (validSecondsLiteral($secondsText)) {
                $seconds = (float) $secondsText;
                if ($tokenStart > $lastEmit) {
                    $parts[] = new SpeechPart('text', substr($text, $lastEmit, $tokenStart - $lastEmit));
                }
                $parts[] = new SpeechPart('pause', seconds: $seconds);
                $lastEmit = $tokenEnd;
                $searchFrom = $tokenEnd;
                continue;
            }
        }
        $searchFrom = $bodyStart;
    }
    if ($lastEmit < strlen($text)) {
        $parts[] = new SpeechPart('text', substr($text, $lastEmit));
    }
    return $parts;
}

function validSecondsLiteral(string $value): bool
{
    if ($value === '') {
        return false;
    }
    $parts = explode('.', $value);
    if (count($parts) > 2) {
        return false;
    }
    foreach ($parts as $part) {
        if ($part === '' || preg_match('/^\d+$/', $part) !== 1) {
            return false;
        }
    }
    return true;
}

File: typecast-swift/Sources/Typecast/SpeechComposer.swift
import Foundation

public enum SpeechPart: Equatable, Sendable {
  case text(String)
  case pause(Double)
}

public struct ComposerSettings: Sendable {
  public var voiceId: String?
  public var model: TTSModel?
  public var language: LanguageCode?
  public var prompt: TTSPrompt?
  public var output: OutputSettings?
  public var seed: Int?

  public init(
    voiceId: String? = nil,
    model: TTSModel? = nil,
    language: LanguageCode? = nil,
    prompt: TTSPrompt? = nil,
    output: OutputSettings? = nil,
    seed: Int? = nil
  ) {
    self.voiceId = voiceId
    self.model = model
    self.language = language
    self.prompt = prompt
    self.output = output
    self.seed = seed
  }
}

private enum ComposerPart {
  case speech(String, ComposerSettings)
  case pause(Double)
}

struct ComposeSegment: Encodable {
  let type: String
  var voiceId: String?
  var text: String?
  var model: TTSModel?
  var language: LanguageCode?
  var prompt: TTSPrompt?
  var output: OutputSettings?
  var seed: Int?
  var durationSeconds: Double?

  enum CodingKeys: String, CodingKey {
    case type, text, model, language, prompt, output, seed
    case voiceId = "voice_id"
    case durationSeconds = "duration_seconds"
  }

  static func tts(_ request: TTSRequest) -> Self {
    Self(
      type: "tts", voiceId: request.voiceId, text: request.text, model: request.model,
      language: request.language, prompt: request.prompt, output: request.output, seed: request.seed
    )
  }

  static func pause(_ seconds: Double) -> Self {
    Self(type: "pause", durationSeconds: seconds)
  }
}

struct ComposeRequest: Encodable {
  let segments: [ComposeSegment]
}

public final class SpeechComposer {
  private let client: TypecastClient
  private var defaultSettings = ComposerSettings()
  private var parts: [ComposerPart] = []

  init(client: TypecastClient) {
    self.client = client
  }

  public func defaults(_ settings: ComposerSettings) -> SpeechComposer {
    defaultSettings = mergeSettings(defaultSettings, settings)
    return self
  }

  public func say(_ text: String, overrides: ComposerSettings = ComposerSettings())
    -> SpeechComposer
  {
    parts.append(.speech(text, mergeSettings(defaultSettings, overrides)))
    return self
  }

  /// Inserts silence between speech segments.
  ///
  /// `seconds` is a duration in seconds. Use `0.3` for 300 ms, `3` for three seconds.
  public func pause(_ seconds: Double) -> SpeechComposer {
    parts.append(.pause(seconds))
    return self
  }

  public func generate() async throws -> TTSResponse {
    let plan = try buildPlan()
    guard
      plan.contains(where: {
        if case .speech = $0 { return true }
        return false
      })
    else {
      throw TypecastError.validationError("at least one speech segment is required")
    }

    let outputFormat = defaultSettings.output?.audioFormat ?? .wav
    var segments: [ComposeSegment] = []
    for part in plan {
      switch part {
      case .pause(let seconds):
        segments.append(.pause(seconds))

      case .speech(let text, let settings):
        segments.append(.tts(try request(text: text, settings: settings, format: outputFormat)))
      }
    }
    return try await client.composeTextToSpeech(segments)
  }

  private func buildPlan() throws -> [ComposerPart] {
    var plan: [ComposerPart] = []
    for part in parts {
      switch part {
      case .pause(let seconds):
        guard seconds.isFinite, seconds > 0 else {
          throw TypecastError.validationError("pause seconds must be greater than 0")
        }
        plan.append(.pause(seconds))

      case .speech(let text, let settings):
        for parsed in parsePauseMarkup(text) {
          switch parsed {
          case .pause(let seconds):
            plan.append(.pause(seconds))
          case .text(let text):
            guard !text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else { continue }
            guard let voiceId = settings.voiceId, !voiceId.isEmpty else {
              throw TypecastError.validationError(
                "voiceId is required for composed speech segments")
            }
            guard settings.model != nil else {
              throw TypecastError.validationError("model is required for composed speech segments")
            }
            plan.append(.speech(text, settings))
          }
        }
      }
    }
    return plan
  }

  private func request(text: String, settings: ComposerSettings, format: AudioFormat) throws
    -> TTSRequest
  {
    return TTSRequest(
      voiceId: settings.voiceId!,
      text: text,
      model: settings.model!,
      language: settings.language,
      prompt: settings.prompt,
      output: mergeOutput(settings.output, OutputSettings(audioFormat: format)),
      seed: settings.seed
    )
  }
}

public func parsePauseMarkup(_ text: String) -> [SpeechPart] {
  var parts: [SpeechPart] = []
  var lastEmit = text.startIndex
  var searchFrom = text.startIndex

  while let startRange = text.range(of: "<|", range: searchFrom..<text.endIndex) {
    let bodyStart = startRange.upperBound
    guard let endRange = text.range(of: "|>", range: bodyStart..<text.endIndex) else {
      break
    }

    let tokenBody = String(text[bodyStart..<endRange.lowerBound])
    if tokenBody.hasSuffix("s") {
      let secondsText = String(tokenBody.dropLast())
      if validSecondsLiteral(secondsText), let seconds = Double(secondsText) {
        if startRange.lowerBound > lastEmit {
          parts.append(.text(String(text[lastEmit..<startRange.lowerBound])))
        }
        parts.append(.pause(seconds))
        lastEmit = endRange.upperBound
        searchFrom = endRange.upperBound
        continue
      }
    }

    searchFrom = bodyStart
  }

  if lastEmit < text.endIndex {
    parts.append(.text(String(text[lastEmit..<text.endIndex])))
  }
  return parts
}

extension TypecastClient {
  public func composeSpeech() -> SpeechComposer {
    SpeechComposer(client: self)
  }
}

private func validSecondsLiteral(_ value: String) -> Bool {
  guard !value.isEmpty else { return false }
  let pieces = value.split(separator: ".", omittingEmptySubsequences: false)
  guard pieces.count <= 2, !pieces[0].isEmpty, pieces[0].allSatisfy(\.isNumber) else {
    return false
  }
  if pieces.count == 2 {
    return !pieces[1].isEmpty && pieces[1].allSatisfy(\.isNumber)
  }
  return true
}

private func mergeSettings(_ base: ComposerSettings, _ override: ComposerSettings)
  -> ComposerSettings
{
  ComposerSettings(
    voiceId: override.voiceId ?? base.voiceId,
    model: override.model ?? base.model,
    language: override.language ?? base.language,
    prompt: override.prompt ?? base.prompt,
    output: mergeOutput(base.output, override.output),
    seed: override.seed ?? base.seed
  )
}

private func mergeOutput(_ base: OutputSettings?, _ override: OutputSettings?) -> OutputSettings? {
  guard base != nil || override != nil else { return nil }
  return OutputSettings(
    volume: override?.volume ?? base?.volume,
    targetLufs: override?.targetLufs ?? base?.targetLufs,
    audioPitch: override?.audioPitch ?? base?.audioPitch,
    audioTempo: override?.audioTempo ?? base?.audioTempo,
    audioFormat: override?.audioFormat ?? base?.audioFormat
  )
}
